### PR TITLE
Add Apple authenticator implementation

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
## Purpose
Sign in with Apple follows the OAuth OIDC flow, but has some deviations from the traditional standards. Hence it requires to implement a separate authenticator extending the OIDC authenticator.

This PR adds the authenticator implementation for the Apple federated identity provider.

## Related Isssue
- https://github.com/wso2/product-is/issues/15501

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+                        .contains("redirect_uri=")) {
+                    authzRequest = OAuthClientRequest.authorizationLocation(authorizationEP).setClientId(clientId)
+                            .setResponseType(OIDCAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE)
+                            .setScope(OIDCAuthenticatorConstants.OAUTH_OIDC_SCOPE).setState(state).buildQueryMessage();
+                } else {
+                    authzRequest = OAuthClientRequest.authorizationLocation(authorizationEP).setClientId(clientId)
+                            .setRedirectURI(callbackUrl)
+                            .setResponseType(OIDCAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE).setScope(scope)
+                            .setState(state).buildQueryMessage();
+                }
```
Comment (wenodpathirana) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:164]:
```suggestion
                OAuthClientRequest.AuthenticationRequestBuilder requestBuilder =
                        OAuthClientRequest.authorizationLocation(authorizationEP).setClientId(clientId)
                                .setResponseType(OIDCAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE).setState(state);

                if (StringUtils.isBlank(queryString) ||
                        !queryString.toLowerCase(Locale.getDefault()).contains("scope=")) {
                    requestBuilder.setScope(scope);
                }
                if (StringUtils.isBlank(queryString) ||
                        !queryString.toLowerCase(Locale.getDefault()).contains("redirect_uri=")) {
                    requestBuilder.setScope(callbackUrl);
                }
                OAuthClientRequest authzRequest = requestBuilder.buildQueryMessage();
```

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,952 @@
...
+     *
+     * @param scope                     Scope.
+     * @param authenticatorProperties   Authenticator properties.
+     * @return Scope string.
+     */
+    @Override
+    protected String getScope(String scope, Map<String, String> authenticatorProperties) {
+
+        if (StringUtils.isBlank(scope)) {
+            return StringUtils.EMPTY;
```
Comment (wenodpathirana) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:484]: If the scope is not set, can't we set the default OIDC scope? 
```suggestion
            return OIDCAuthenticatorConstants.OAUTH_OIDC_SCOPE;
```

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,952 @@
...
+        }
+
```
Comment (wenodpathirana) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:486]:
```suggestion
```

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+                        currentEpochTime, expiryEpochTime);
+                log.debug(msg);
+            }
+
+            // Check for pre-conditions.
+            if (StringUtils.isBlank(authenticatorProperties.get(AppleAuthenticatorConstants.PRIVATE_KEY))
+                    || StringUtils.isBlank(authenticatorProperties.get(AppleAuthenticatorConstants.TEAM_ID))
+                    || StringUtils.isBlank(authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_ID))
+                    || StringUtils.isBlank(authenticatorProperties.get(AppleAuthenticatorConstants.KEY_ID))
+            ) {
```
Comment (wenodpathirana) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:626]: Shouldn't these be mandatory?

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java (hunk 2 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:484]: Apple doesn't require sending `openid` scope. Hence it doesn't mean anything.

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java (hunk 4 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:626]: Yes. These attributes are set as required fields in [authenticator configs](https://github.com/wso2-extensions/identity-outbound-auth-apple/pull/1/files#diff-7d561c19910ba6eb5ad3e88a271822acdd95c67bfc6dda4cb96e2a595a12c99fR343). However it is possible to set them empty through the API/ management console.
Therefore I have added a check here.

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java (hunk 1 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: Fixed with 1b7d55fadaf1055ede63cc314115132750888db2

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java (hunk 3 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: Fixed with 1b7d55fadaf1055ede63cc314115132750888db2

**Review by wenodpathirana — DISMISSED** [review]

**Review by wenodpathirana — APPROVED** [review]

**Pull request merged**